pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -205,59 +205,49 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 		gc.workerLock.Lock()
 		defer gc.workerLock.Unlock()
```

Comment:
I wonder if we can drop this worker lock.

cc @deads2k 

This PR changes the GC sync loop to unlock workers and continue running them even if the sync fails. That's only ok if we consider running workers with unsynced informers to be safe.

If we consider that to be safe, do we even need to lock workers in Sync at all? The main scenario I could think of where that would be beneficial would be a happy path first-sync case where it saves workers from a lot of virtual-node bookkeeping if all normal known types generally have synced informers. But my read of the Run method above is that we wait for an initial sync before even starting workers the first time, so I don't think that applies.

Reply:
> my read of the Run method above is that we wait for an initial sync before even starting workers the first time

speaking of which, if we intend for GC to be able to run with unavailable / unhealthy resources, I think this bit in Run also has to change from an unconditional wait to a timeout:

```
	if !cache.WaitForNamedCacheSync("garbage collector", ctx.Done(), gc.dependencyGraphBuilder.IsSynced) {
		return
	}
```

Reply:
> But my read of the Run method above is that we wait for an initial sync before even starting workers the first time, so I don't think that applies.

This is true, except for in the case that this PR tries to address (where at least one informer is unable to sync). In those cases, we _will_ have unsynced informers. Though others are likely to have synced successfully, meaning progress can still be made for those types at least.

I'm not so familiar with whether virtual-node bookkeeping is problematic or when/where it'd be triggered in relation to this change..

What's the best way to verify whether removing this lock is a problem? It will certainly require some tests to be amended/removed.. can we proceed as-is, and revisit the lock after?